app/src/main/java/org/oppia/app/parser/StringToFractionParser.kt

```diff
@@ -8,14 +12,55 @@ class StringToFractionParser {
   private val wholeNumberOnlyRegex = """^-? ?(\d+)$""".toRegex()
   private val fractionOnlyRegex = """^-? ?(\d+) ?/ ?(\d+)$""".toRegex()
   private val mixedNumberRegex = """^-? ?(\d+) (\d+) ?/ ?(\d+)$""".toRegex()
+  private val invalidCharsRegex = """^[\d\s/-]+$""".toRegex()
 
-  fun getFractionFromString(text: String): Fraction {
+  /**
+   * Returns a [FractionParsingError] for the specified text input if it's an invalid fraction, or
+   * [FractionParsingError.VALID] if no issues are found. Note that a valid fraction returned by this method is guaranteed
+   * to be parsed correctly by [parseRegularFraction].
+   *
+   * This method should only be used when a user tries submitting an answer. Real-time error detection should be done
+   * using [getPendingAnswerErrorOnRealTime], instead.
+   */
+  fun getSubmitTimeError(text: String): FractionParsingError {
+    val fraction = parseFraction(text)
+    return when {
+      fraction == null -> FractionParsingError.INVALID_FORMAT
+      fraction.denominator == 0 -> FractionParsingError.DIVISION_BY_ZERO
+      else -> FractionParsingError.VALID
+    }
+  }
+
+  /**
+   * Returns a [FractionParsingError] for obvious incorrect fraction formatting issues for the specified raw text, or
+   * [FractionParsingError.VALID] if not such issues are found.
+   *
+   * Note that this method returning a valid result does not guarantee the text is a valid fraction--
+   * [getSubmitTimeError] should be used for that, instead. This method is meant to be used as a quick sanity check for
+   * general validity, not for definite correctness.
+   */
+  fun getPendingAnswerErrorOnRealTime(text: String): FractionParsingError {
+    val normalized = text.normalizeWhitespace()
+    return when {
+      !normalized.matches(invalidCharsRegex) -> FractionParsingError.INVALID_CHARS
+      normalized.startsWith("/") -> FractionParsingError.INVALID_FORMAT
+      normalized.count { it == '/' } > 1 -> FractionParsingError.INVALID_FORMAT
+      normalized.lastIndexOf('-') > 0 -> FractionParsingError.INVALID_FORMAT
+      else -> FractionParsingError.VALID
+    }
+  }
+
+  private fun parseFraction(text: String): Fraction? {
     // Normalize whitespace to ensure that answer follows a simpler subset of possible patterns.
     val inputText: String = text.normalizeWhitespace()
     return parseMixedNumber(inputText)
-      ?: parseFraction(inputText)
+      ?: parseRegularFraction(inputText)
       ?: parseWholeNumber(inputText)
-      ?: throw IllegalArgumentException("Incorrectly formatted fraction: $text")
+      ?: null
+  }
+
+  fun getFractionFromString(text: String): Fraction {
```

Comment:
Please add a brief KDoc for this method since it's part of the public API of this class, e.g.:

```kotlin
/** Returns a [Fraction] parse from the specified raw text string. */
```

Reply:
Added Kdoc